src/ape/api/transactions.py

```diff
@@ -230,21 +230,29 @@ def _confirmations_occurred(self) -> int:
 
         return latest_block.number - self.block_number
 
+    @cached_property
+    def events(self) -> List[ContractLog]:
```

Comment:
What is the reason this has been made a list (a mutable object)? Do we really require mutability here?
My feeling is this should have been a tuple, an immutable object.

Reply:
Oh ya, I guess I meant sorted, not sort.

Reply:
> sorted_logs = list(tx.events).sort(key=foo)

This wouldn't work, but this would:
```python
sorted_logs = sorted(tx.events, key=foo)
```

Reply:
However! I just noticed that `sorted()` can take a tuple, but it returns a list.
Still I am leaning `list` though.

Reply:
> Still I am leaning list though.

Let me elaborate on why I think it is a good idea to go with an immutable type in cases like this.
If it is not immediately obvious, perhaps by looking at the function's name, that the function returns a new object,
without references being held by the original object (e.g . the instance in case of instance methods), then one
may think that either (1) the return value is not referenced by the original object and is therefore safe to mutate, or (2) the return value _is_ referenced by the original object and is therefore not safe to mutate. Either of those would be valid. This situation is even more pronounced in cases like this where we have an attribute access.

By choosing immutable types over mutable ones, the API authors signal the correct assumption to the user of the API.
For that reason it also a fairly wide practice in the Python community.

Reply:
Since it is a cached_property, it is technically is referenced by the original object, but I see what you mean.

Reply:
> Since it is a cached_property, it is technically is referenced by the original object

... which means that repeated attribute accesses of `.events` on the same object are not guaranteed to return the same set of event, right?

Reply:
I think this is a good catch, and it should have been a normal `property` (consistent with how `ContractEvent.from_receipt` is implemented) and produce a new list object every time it's called.

It adds a small amount of overhead, but there are no RPC calls done in response to triggering this property, just a call to `EcosystemAPI` for decoding the log data with the given ABIs.

Not fundamentally against tuples, as it's only slightly more awkward to remember the comma when comparing against expected logs in a test, but I think a list is pretty natural to work with, and you can rearrange it like @unparalleled-js mentioned to have an easier time comparing e.g. `assert sorted(tx.events, key=event_sorting_fn)[:2] == [EventA(...), EventB(...)] ` (has at least two events in any order with those exact comparisons)

Reply:
@cached_property has been removed. Thanks for patience and detailed explanation @istankovic.

Reply:
> ... which means that repeated attribute accesses of `.events` on the same object are not guaranteed to return the same set of events, right?

`@cached_property` does return the same set each time because it is cached on the object.

But now, we made it a normal calculated property. So you can mutate it without affects the receipt.